opentelemetry-sdk/src/opentelemetry/sdk/metrics/__init__.py

```diff
@@ -0,0 +1,307 @@
+
+    def __init__(
+        self,
+        name: str,
+        description: str,
+        unit: str,
+        value_type: metrics_api.ValueType,
+        label_keys: Tuple[str, ...] = None,
+        enabled: bool = True,
+        monotonic: bool = False,
+    ):
+        self.name = name
+        self.description = description
+        self.unit = (unit,)
+        self.value_type = value_type
+        self.label_keys = label_keys
+        self.enabled = enabled
+        self.monotonic = monotonic
+        self.handles = {}
+
+    def get_handle(
+        self, label_values: Tuple[str, ...]
+    ) -> metrics_api.MetricHandle:
+        """See `opentelemetry.metrics.Metric.get_handle`."""
+        pass  # pylint: disable=unnecessary-pass
+
+    def remove_handle(self, label_values: Tuple[str, ...]) -> None:
+        """See `opentelemetry.metrics.Metric.remove_handle`."""
+        self.handles.pop(label_values, None)
+
+    def clear(self) -> None:
+        """See `opentelemetry.metrics.Metric.clear`."""
+        self.handles.clear()
+
+
+class Counter(Metric):
+    """See `opentelemetry.metrics.Counter`.
+
+    By default, counter values can only go up (monotonic). Negative inputs
+    will be discarded for monotonic counter metrics. Counter metrics that
+    have a monotonic option set to False allows negative inputs.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        description: str,
+        unit: str,
+        value_type: metrics_api.ValueType,
+        label_keys: Tuple[str, ...] = None,
+        enabled: bool = True,
+        monotonic: bool = True,
+    ):
+        super(Counter, self).__init__(
+            name,
+            description,
+            unit,
+            value_type,
+            label_keys=label_keys,
+            enabled=enabled,
+            monotonic=monotonic,
+        )
+
+    def get_handle(
+        self, label_values: Tuple[str, ...]
+    ) -> metrics_api.CounterHandle:
+        """See `opentelemetry.metrics.FloatCounter.get_handle`."""
+        handle = self.handles.get(
+            label_values,
+            CounterHandle(self.value_type, self.enabled, self.monotonic),
+        )
+        self.handles[label_values] = handle
+        return handle
+
+
+class Gauge(Metric):
+    """See `opentelemetry.metrics.Gauge`.
+
+    By default, gauge values can go both up and down (non-monotonic).
+    Negative inputs will be discarded for monotonic gauge metrics.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        description: str,
+        unit: str,
+        value_type: metrics_api.ValueType,
+        label_keys: Tuple[str, ...] = None,
+        enabled: bool = True,
+        monotonic: bool = False,
+    ):
+        super(Gauge, self).__init__(
+            name,
+            description,
+            unit,
+            value_type,
+            label_keys=label_keys,
+            enabled=enabled,
+            monotonic=monotonic,
+        )
+
+    def get_handle(
```

Comment:
The `get_handle` implementations are all the same except for the type. So you could it them in the base class and just replace e.g. `GaugeHandle` with `self.handle_type` which you would define as a static constant in derived classes.

Reply:
Just to clarify, are you suggesting to move the `get_handle` logic to the base `Metric` class? If so:
If `self.handle_type` is defined as constants in the derived classes (`Counter`, `Gauge`, `Measure`), the `get_handle` implementation, if moved to the base `Metric` class, won't be able to see the type. 
As a solution, the `handle_type` declaration can be moved to the `Metric` class, with a default value (probably the DefaultMetricHandle` from the API), and then the derived metric class (`Counter`, `Gauge`, `Measure`) will each have their own `handle_type`, overriding the default.

Reply:
That sounds right. If I'm not missing something, the default value will only be "used" to keep the lint (pylint, mypy) silent though. I think, alternatively you could make the SDK Metric base class an ABC and instead of handle_type being a constant it could be an abstract method.